### PR TITLE
Add worker scaling into EvalAI host UI

### PR DESCRIPTION
**Pull request by matthew-so**
IMPORTANT NOTES (please read, then delete):

* The PR title should start with "Fix #bugnum: " (if applicable), followed by a clear one-line present-tense summary of the changes introduced in the PR. For example: "Fix #bugnum: Introduce the first version of the collection editor.".

* Please make sure to mention "#bugnum" somewhere in the description of the PR. This enables GitHub to link the PR to the corresponding bug.

Please also make sure to follow the [style rules](https://github.com/Cloud-CV/EvalAI/blob/master/.github/CONTRIBUTING.md#style-rules).

### CODE REVIEW

**Comment by matthew-so** [comment]
Note that there are some spacing issues in aws_utils.py. This will be fixed ASAP.

File: apps/challenges/views.py [hunk 1]
```diff
@@ -2969,6 +2970,66 @@ def get_worker_logs(request, challenge_pk):
...
+@throttle_classes([UserRateThrottle])
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((JWTAuthentication, ExpiringTokenAuthentication))
+def scale_resources_by_challenge_pk(request, challenge_pk):
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
+        response_data = {
+            "error": "Sorry, you are not authorized for access worker operations."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+    if request.data.get("cores") is None:
```
Comment (Ram81) [inline apps/challenges/views.py:2983]: Please add a new line above this line

Comment (Ram81) [inline apps/challenges/views.py:2983]: Please add a new line above this line

File: apps/challenges/views.py [hunk 2]
```diff
@@ -2969,6 +2970,66 @@ def get_worker_logs(request, challenge_pk):
...
+        response_data = {
+            "error": "Number of cores is missing from request."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+    if request.data.get("memory") is None:
```
Comment (Ram81) [inline apps/challenges/views.py:2988]: same here

File: apps/challenges/views.py [hunk 3]
```diff
@@ -2969,6 +2970,66 @@ def get_worker_logs(request, challenge_pk):
     return Response(response_data, status=status.HTTP_200_OK)
 
 
+@api_view(["PUT"])
+@throttle_classes([UserRateThrottle])
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((JWTAuthentication, ExpiringTokenAuthentication))
+def scale_resources_by_challenge_pk(request, challenge_pk):
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
```
Comment (Ram81) [inline apps/challenges/views.py:2978]: Please add appropriate doc strings. See `scale_workers` method for a example

File: apps/challenges/views.py [hunk 4]
```diff
@@ -2969,6 +2970,66 @@ def get_worker_logs(request, challenge_pk):
...
+        response_data = {
+            "error": "Number of cores is missing from request."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+    if request.data.get("memory") is None:
+        response_data = {
+            "error": "Amount of memory is missing from request."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+    challenge = get_challenge_model(challenge_pk)
```
Comment (Ram81) [inline apps/challenges/views.py:2993]: please add a newline above this line

File: apps/challenges/views.py [hunk 5]
```diff
@@ -2969,6 +2970,66 @@ def get_worker_logs(request, challenge_pk):
...
+    cores = request.data["cores"]
+    memory = request.data["memory"]
+    try:
```
Comment (Ram81) [inline apps/challenges/views.py:2996]: newline above this line. See other methods for reference on code style

File: apps/challenges/views.py [hunk 6]
```diff
@@ -2969,6 +2970,66 @@ def get_worker_logs(request, challenge_pk):
...
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    # validate cores and memory number
+    if (
+        cores == 256 and memory in (512, 1024, 2048)
+        or cores == 512 and memory in (1024, 2048, 3072, 4096)
+        or cores == 1024 and memory in (2048, 3072, 4096, 5120, 6144, 7168, 8192)
+        or cores == 2048 and memory >= 4096 and memory <= 16384 and memory % 1024 == 0
+        or cores == 4096 and memory >= 8192 and memory <= 30720 and memory % 1024 == 0
```
Comment (Ram81) [inline apps/challenges/views.py:3011]: For now can we only allow scaling to `1024` vcpu and `2048` memory maximum

File: apps/challenges/aws_utils.py [hunk 7]
```diff
@@ -95,6 +95,96 @@
 }
 
 
+def scale_resources(challenge, new_cores, new_memory):
+    client = get_boto3_client("ecs", aws_keys)
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:99]: @matthew-so  take a look at other fargate worker util methods, for example, `scale_workers` method [here](https://github.com/Cloud-CV/EvalAI/blob/master/apps/challenges/aws_utils.py#L564). Please follow similar docstring and first set of checks. We only want to be able to call the API in prod and staging and not locally.

File: apps/challenges/aws_utils.py [hunk 8]
```diff
@@ -95,6 +95,96 @@
...
+    if challenge.worker_cpu_cores == new_cores and challenge.worker_memory == new_memory:
+        message = "Error. Worker cores and memory were not modified."
+        return {
+            "Error": message,
+            "ResponseMetadata": {"HTTPStatusCode": HTTPStatus.BAD_REQUEST},
+        }
+    if not challenge.task_def_arn:
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:106]: please add a newline above it

File: apps/challenges/aws_utils.py [hunk 9]
```diff
@@ -95,6 +95,96 @@
...
+        message = "Error. Task definition not yet registered for challenge{}.".format(
+            challenge.pk
+        )
+        return {
+            "Error": message,
+            "ResponseMetadata": {"HTTPStatusCode": HTTPStatus.BAD_REQUEST},
+        }
+    # deregister
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:114]: Can you make the comment more descriptive? Please take a look at other APIs for reference

File: apps/challenges/aws_utils.py [hunk 10]
```diff
@@ -95,6 +95,96 @@
...
+    try:
+        response = client.deregister_task_definition(
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:116]: Deregistering a task def will take a couple of minutes and you won't be able to create a new service with similar task def that has same service name. We have to make this a celery task and call the celery task from the API which will sync asynchronously. Please read about celery tasks [here](https://docs.celeryq.dev/en/stable/userguide/tasks.html)

File: apps/challenges/aws_utils.py [hunk 11]
```diff
@@ -270,8 +360,8 @@ def register_task_def_by_challenge_pk(client, queue_name, challenge):
             try:
                 response = client.register_task_definition(**definition)
                 if (
-                    response["ResponseMetadata"]["HTTPStatusCode"]
-                    == HTTPStatus.OK
+                        response["ResponseMetadata"]["HTTPStatusCode"]
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:363]: can you undo the formatting changes?

**Review by Ram81 — CHANGES_REQUESTED** [review]

**Comment by matthew-so** [comment]
The change is complete. The total deregistration and re-registration process takes less than two seconds in my testing, making it fast enough to run synchronously. One potential improvement is changing the CPU and memory fields into a dropdown box in order to make the list of valid configurations clearer. Please let me know if any other changes would be helpful.

Screenshot 2022-09-19 013946

File: apps/challenges/aws_utils.py [hunk 12]
```diff
@@ -615,6 +616,115 @@ def scale_workers(queryset, num_of_tasks):
     return {"count": count, "failures": failures}
 
 
+def scale_resources(challenge, new_cpu_units, new_memory):
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:619]: can you update the variable names to `worker_cpu_cores` and `worker_memory`

File: apps/challenges/aws_utils.py [hunk 13]
```diff
@@ -615,6 +616,115 @@ def scale_workers(queryset, num_of_tasks):
...
+    if not challenge.task_def_arn:
+        message = "Error. Task definition not yet registered for challenge{}.".format(
+            challenge.pk
+        )
+        return {
+            "Error": message,
+            "ResponseMetadata": {"HTTPStatusCode": HTTPStatus.BAD_REQUEST},
+        }
+
+    # deregister
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:656]: Change comment to `Deregister current worker service task definition`

File: apps/challenges/views.py [hunk 14]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+    Returns:
+        Response object -- Response object with appropriate response code (200/400/403/404)
+    """
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
+        response_data = {
+            "error": "Sorry, you are not authorized for access worker operations."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    if request.data.get("cpu_units") is None:
```
Comment (Ram81) [inline apps/challenges/views.py:2997]: Can we rename the key to `worker_cpu_cores`

File: apps/challenges/views.py [hunk 15]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+        response_data = {
+            "error": "Number of CPU units is missing from request."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    if request.data.get("memory") is None:
```
Comment (Ram81) [inline apps/challenges/views.py:3003]: can we rename the key to `worker_memory`?

File: apps/challenges/views.py [hunk 16]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+        response_data = {
+            "error": "Amount of memory is missing from request."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    challenge = get_challenge_model(challenge_pk)
+    cpu_units = request.data["cpu_units"]
+    memory = request.data["memory"]
```
Comment (Ram81) [inline apps/challenges/views.py:3011]: same here

File: apps/challenges/views.py [hunk 17]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+
+    try:
+        cpu_units = int(cpu_units)
+        memory = int(memory)
+    except Exception:
+        response_data = {
+            "error": "Enter integer numbers for cores and memory."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
```
Comment (Ram81) [inline apps/challenges/views.py:3020]: Is there a cleaner way to write this check? A separate try except just for parsing seems bad idea

File: apps/challenges/views.py [hunk 18]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+
+    # validate cores and memory number
+    if (
+        cpu_units == 256 and memory in (512, 1024, 2048)
+        or cpu_units == 512 and memory in (1024, 2048)
+        or cpu_units == 1024 and memory == 2048
+    ):
+        challenge.queue = "random-number-generator-challenge-596-staging-55d46e9a-563d-48f1-a4d7-b9a7f36a55"
+        challenge.task_def_arn = "random-number-generator-challenge-596-staging-55d46e9a-563d-48f1-a4d7-b9a7f36a55:4"
```
Comment (Ram81) [inline apps/challenges/views.py:3029]: Please remove these before deployment

File: apps/challenges/views.py [hunk 19]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+            return Response(response_data, status=status.HTTP_503_SERVICE_UNAVAILABLE)
+        else:
+            response_data = {
+                "Success": "The challenge {} has been scaled successfully".format(
+                    challenge.title
+                )
+            }
+    else:
+        response_data = {
+            "error": "Invalid core and memory combination."
```
Comment (Ram81) [inline apps/challenges/views.py:3044]: Change message to `Please specify correct worker CPU cores and memory config.`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 20]
```diff
@@ -1620,6 +1623,36 @@
             utilities.sendRequest(parameters);
         };
 
+        vm.setWorkerResources = function() {
+            parameters.url = "challenges/"+vm.challengeId+"/scale_resources/";
+            parameters.method = 'PUT';
+            parameters.data = {
+                "cpu_units": vm.cpuUnits,
+                "memory": vm.memory,
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:1631]: can we rename the keys to `worker_cpu_cores` and `worker_memory`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 21]
```diff
@@ -1620,6 +1623,36 @@
             utilities.sendRequest(parameters);
         };
 
+        vm.setWorkerResources = function() {
+            parameters.url = "challenges/"+vm.challengeId+"/scale_resources/";
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:1627]: Please add spaces after and before `+`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 22]
```diff
@@ -1620,6 +1623,36 @@
...
+            parameters.method = 'PUT';
+            parameters.data = {
+                "cpu_units": vm.cpuUnits,
+                "memory": vm.memory,
+            };
+            parameters.callback = {
+                onSuccess: function() {
+                    $rootScope.notify("success", "Worker resources have been changed successfully!");
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:1635]: Change the toast message to `Evaluation worker scaled successfully`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 23]
```diff
@@ -1620,6 +1623,36 @@
...
+                onError: function(response) {
+                    vm.team.error = true;
+                    vm.stopLoader();
+                    let requestError = '';
+                    if (typeof(response.data) == 'string') {
+                        requestError = response.data;
+                    } else {
+                        requestError = JSON.stringify(response.data, null, 4);
+                    }
+                    $rootScope.notify("error", "Worker resources could not be changed: " + requestError);
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:1649]: Change the toast message to `Error scaling evaluation worker: {}`

File: apps/challenges/views.py [hunk 24]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+
+    if request.data.get("cpu_units") is None:
+        response_data = {
+            "error": "Number of CPU units is missing from request."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    if request.data.get("memory") is None:
+        response_data = {
+            "error": "Amount of memory is missing from request."
```
Comment (Ram81) [inline apps/challenges/views.py:3005]: change message to `Worker memory config missing from request data`

File: apps/challenges/views.py [hunk 25]
```diff
@@ -2969,6 +2970,83 @@ def get_worker_logs(request, challenge_pk):
...
+    """
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
+        response_data = {
+            "error": "Sorry, you are not authorized for access worker operations."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    if request.data.get("cpu_units") is None:
+        response_data = {
+            "error": "Number of CPU units is missing from request."
```
Comment (Ram81) [inline apps/challenges/views.py:2999]: Change mssage to `Number of CPU units config missing from request data`

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: apps/challenges/aws_utils.py (hunk 12 above)
Comment (matthew-so) [reply apps/challenges/aws_utils.py:619]: For vCPU count, why are they currently named "cpu_cores", considering that the vCPU count given to AWS does not correspond to actual virtual or physical cores available to the worker?

File: apps/challenges/views.py (hunk 18 above)
Comment (matthew-so) [reply apps/challenges/views.py:3029]: Sounds good! Removing it now.

File: apps/challenges/views.py (hunk 17 above)
Comment (matthew-so) [reply apps/challenges/views.py:3020]: This is the cleanest way to convert the entries into ints, as is shown [here](https://stackoverflow.com/questions/736043/checking-if-a-string-can-be-converted-to-float-in-python).

File: apps/challenges/aws_utils.py [hunk 26]
```diff
@@ -615,6 +615,115 @@ def scale_workers(queryset, num_of_tasks):
...
+        ):
+            challenge.task_def_arn = None
+            challenge.save()
+        else:
+            return response
+    except ClientError as e:
+        logger.exception(e)
+        return e.response
+
+    # register
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:672]: remove the comment

File: apps/challenges/aws_utils.py [hunk 27]
```diff
@@ -615,6 +615,115 @@ def scale_workers(queryset, num_of_tasks):
...
+        AWS_SES_REGION_NAME=AWS_SES_REGION_NAME,
+        AWS_SES_REGION_ENDPOINT=AWS_SES_REGION_ENDPOINT,
+        **COMMON_SETTINGS_DICT,
+        **challenge_aws_keys,
+    )
+    definition = eval(definition)
+
+    try:
+        response = client.register_task_definition(**definition)
+        # if we successfully register a new task definition, update the service
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:697]: remove the comment

File: frontend/src/js/controllers/challengeCtrl.js [hunk 28]
```diff
@@ -97,6 +97,9 @@
 
         vm.workerLogs = [];
 
+        vm.cpuCores = 256;
+        vm.memory = 512;
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:101]: Can we update these to:
```
vm.cpuCores = 512
vm.memory = 1024
```

Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:101]: Can you also rename these variables to `workerMemory` and `workerCPUCores`

**Review by Ram81 — CHANGES_REQUESTED** [review]
@matthew-so minor changes, everything else looks good. Can you also share a screenshot of the UI?

**Comment by Ram81** [comment]
@matthew-so can we move these dropdowns and scale button above the `start/stop/restart` buttons. Also, can you show both CPU cores and worker memory dropdowns next to each other and both dropdown components should have a smaller width.

**Review by Ram81 — CHANGES_REQUESTED** [review]
@matthew-so can you push the latest changes and share the screenshot of the manage tab UI after changes we discussed?

**Comment by matthew-so** [comment]
new_worker_ui

**Review by Ram81 — APPROVED** [review]
LGTM :+1:

**Pull request merged**